Port the `#[linkage]` attribute to the new attribute system

scrabsha opened the pull request:
<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

r? @jdonszelmann

scrabsha committed:
Port the `#[linkage]` attribute to the new attribute system

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16918600484/job/47938797072) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/47938797072) [(plain)](https://github.com/rust-lang/rust/commit/4e422232909af4749b0a5bdc1475a3f3c3fd4109/checks/47938797072/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
tidy check
Checking tidy rustdoc_json...
No error code explanation was removed!
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/compiler/rustc_hir/src/attrs/encode_cross_crate.rs:47: line not in alphabetical order
##[error]tidy error: /checkout/compiler/rustc_hir/src/attrs/data_structures.rs:382: line not in alphabetical order
##[error]tidy error: /checkout/compiler/rustc_span/src/symbol.rs:1286: line not in alphabetical order
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'virtualenv'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (25.1.1)
Collecting pip
---

added 291 packages in 37s
Running eslint on rustdoc JS files
info: ES-Check: there were no ES version matching errors!  🎉
some tidy checks failed
Command `/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rust-tidy /checkout /checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo /checkout/obj/build 4 /node/bin/npm --extra-checks=py,cpp,js` failed with exit code 1
Created at: src/bootstrap/src/core/build_steps/tool.rs:1636:23
Executed at: src/bootstrap/src/core/build_steps/test.rs:1239:29

Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:01:56
  local time: Tue Aug 12 19:23:17 UTC 2025
  network time: Tue, 12 Aug 2025 19:23:17 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

jdonszelmann commented:
@rustbot author

scrabsha commented:
I removed `rustc_middle::mir::Linkage` and replaced its usage by `rustc_hir::attrs::Linkage`.

I also renamed some variants of `rustc_hir::attrs::Linkage` to match the names that were used in `rustc_midle::mir::Linkage` and in the LLVM API.

@rustbot review

jdonszelmann commented:
@bors r+ rollup

bors commented:
:pushpin: Commit d435197afcb019a692aa1faf8b7169e167ac1de8 has been approved by `jdonszelmann`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jdonszelmann d435197afcb019a692aa1faf8b7169e167ac1de8 -->
<!-- homu: {"type":"Approved","sha":"d435197afcb019a692aa1faf8b7169e167ac1de8","approver":"jdonszelmann","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
